### PR TITLE
[MemProf] Remove empty edges once after cloning

### PR DESCRIPTION
**Pull request by teresajohnson**
Restructure the handling of edges that become empty during the cloning
process. Instead of removing them as they become empty (no context ids
and alloc type), do this once after all cloning is complete.

This has no effect on the cloning result, but prepares for a follow on
change that does improve the cloning. The structural change here reduces
the diffs for the follow on change, which would be much more difficult
with the previous handling.

### CODE REVIEW

File: llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp [hunk 1]
```diff
@@ -306,7 +306,7 @@ class CallsiteContextGraph {
     // True if this node was effectively removed from the graph, in which case
     // its context id set, caller edges, and callee edges should all be empty.
     bool isRemoved() const {
-      assert(ContextIds.empty() ==
+      assert(!ContextIds.empty() ||
```
Comment (snehasish) [inline llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:309]: An assert with an or condition is a little hard to reason about for me. For this to fail both LHS and RHS need to be false - context IDs is empty *and* at least one of CalleeEdges, CallerEdges is not empty. Can we rewrite this as two separate assertions to simplify? Also it seems that it is now inconsistent with the comment above?

**Review by snehasish — COMMENTED** [review]
Does it make sense to add a separate test for this or is it covered in upcoming changes?

File: llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp (hunk 1 above)
Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:309]: The original assert was wrong, which showed up in testing after the reorganization here. The case that was wrong was when we were left with a single (non-removed) node in the graph, which would have non-empty context ids, but no callee or caller edges. I played around with restructuring the assertion but didn't come up with anything that looked clearer to me. E.g. here's another structure but it feels more convoluted to me personally:

```
assert(!(ContextIds.empty() &&
               (!CalleeEdges.empty() || !CallerEdges.empty()));
```

It's not inconsistent with the comment (it is still true that all 3 should be empty if it was removed) but I've added new comments to explain the case described above. Let me know if you have a preference of what to do for the assert structure.

**Comment by teresajohnson** [comment]
> Does it make sense to add a separate test for this or is it covered in upcoming changes?

This is essentially an NFC change (I thought about adding NFC to it but decided not since it changes enough about the internal ordering of the graph transformation, although it should have no impact on anything user-visible).

**Review by snehasish — APPROVED** [review]
lgtm

File: llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp (hunk 1 above)
Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:311]: It is equivalent, it just means an extra check that is executed unnecessarily. Although hopefully the compiler would common it with the check on the return?

Comment (snehasish) [reply llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:311]: I dont think there will be an extra check. With clang for non-assert builds (where NDEBUG is defined) the IR has an empty basic block [1] which should get optimized away in the generated code.

[1] https://godbolt.org/z/r5fceqeTK

Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:311]: Good point, will go ahead and restructure like that.

**Pull request merged**